### PR TITLE
feat: autotracking for reference id access

### LINKED ISSUE #7102: Reference API is currently incompatible with Tracked
If a getter uses the references API since no leaves are tracked it will be memoized by the template and never re-run. To fix this our previously discussed creation of per-record tags for properties to consume should be completed and reference methods should consume those tags.

### PR DESCRIPTION
**Pull request by runspired**
resolves #7102

### CODE REVIEW

File: packages/store/addon/-private/system/references/belongs-to.ts [hunk 1]
```diff
@@ -22,17 +30,76 @@ import Reference from './reference';
...
-  constructor(store, parentIMOrIdentifier, belongsToRelationship, key) {
-    super(store, parentIMOrIdentifier);
+  declare key: string;
+  declare belongsToRelationship: BelongsToRelationship;
+  declare type: string;
+  declare parent: RecordReference;
+  declare parentIdentifier: StableRecordIdentifier;
+
+  #token: Object;
+  #relatedToken: Object | null = null;
```
Comment (snewcomer) [inline packages/store/addon/-private/system/references/belongs-to.ts:40]: Can we put an inline comment on what these are used for?

File: packages/-ember-data/tests/integration/references/autotracking-test.js [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+    await render(hbs`id: {{if this.context.id this.context.id 'null'}}`);
+
+    assert.strictEqual(getRootElement().textContent, 'id: null', 'the id is null');
+    assert.strictEqual(testContext.id, null, 'the id is correct initially');
+    await dan.save();
+    await settled();
+    assert.strictEqual(getRootElement().textContent, 'id: 6', 'the id updates when the record id updates');
+    assert.strictEqual(testContext.id, '6', 'the id is correct when the record is saved');
+  });
+});
```
Comment (snewcomer) [inline packages/-ember-data/tests/integration/references/autotracking-test.js:151]: 👍

File: packages/store/addon/-private/system/references/belongs-to.ts [hunk 3]
```diff
@@ -22,17 +30,76 @@ import Reference from './reference';
...
+  get _relatedIdentifier(): StableRecordIdentifier | null {
+    this._ref; // consume the tracked prop
+    if (this.#relatedToken) {
+      unsubscribe(this.#relatedToken);
+    }
+
+    let resource = this._resource();
+    if (resource && resource.data) {
+      const identifier = this.store.identifierCache.getOrCreateRecordIdentifier(resource.data);
+      this.#relatedToken = this.store._notificationManager.subscribe(
```
Comment (snewcomer) [inline packages/store/addon/-private/system/references/belongs-to.ts:88]: The reference API is for `CUSTOM_MODEL_CLASS` true and false right?  Do subscription need to be protected if `CUSTOM_MODEL_CLASS` is turned off?

Comment (runspired) [reply packages/store/addon/-private/system/references/belongs-to.ts:88]: that would be the test failure 🤦🏻

**Pull request merged**